fluss-server/src/main/java/org/apache/fluss/server/kv/rowmerger/AggregateRowMerger.java

```diff
@@ -0,0 +1,399 @@
+        this.deleteBehavior = tableConfig.getDeleteBehavior().orElse(DeleteBehavior.IGNORE);
+        this.contextCache = new AggregationContextCache(schemaGetter, kvFormat);
+        // Initialize cache with same settings as PartialUpdaterCache and AggregationContextCache
+        this.partialMergerCache =
+                Caffeine.newBuilder()
+                        .maximumSize(PARTIAL_MERGER_CACHE_MAX_SIZE)
+                        .expireAfterAccess(PARTIAL_MERGER_CACHE_EXPIRE_DURATION)
+                        .build();
+    }
+
+    @Override
+    public BinaryValue merge(BinaryValue oldValue, BinaryValue newValue) {
+        // First write: no existing row
+        if (oldValue == null || oldValue.row == null) {
+            return newValue;
+        }
+
+        // Get contexts for schema evolution support
+        // Optimize: reuse context when schemaId is the same, and avoid Tuple2 allocation
+        AggregationContext oldContext;
+        AggregationContext newContext;
+
+        if (oldValue.schemaId == newValue.schemaId) {
+            // Same schema: reuse the same context to avoid duplicate lookup
+            newContext = contextCache.getContext(newValue.schemaId);
+            oldContext = newContext;
+        } else {
+            // Different schemas: need separate lookups
+            oldContext = contextCache.getContext(oldValue.schemaId);
+            newContext = contextCache.getContext(newValue.schemaId);
+        }
+
+        // Use row encoder for aggregation
+        RowEncoder encoder = newContext.getRowEncoder();
+        encoder.startNewRow();
+        AggregateFieldsProcessor.aggregateAllFields(
+                oldValue.row, newValue.row, oldContext, newContext, encoder);
+        BinaryRow mergedRow = encoder.finishRow();
+
+        return new BinaryValue(newValue.schemaId, mergedRow);
+    }
+
+    @Override
+    public BinaryValue delete(BinaryValue oldValue) {
+        // Remove the entire row (returns null to indicate deletion)
+        return null;
+    }
+
+    @Override
+    public DeleteBehavior deleteBehavior() {
+        return deleteBehavior;
+    }
+
+    @Override
+    public RowMerger configureTargetColumns(
+            @Nullable int[] targetColumns, short schemaId, Schema schema) {
+        if (targetColumns == null) {
+            return this;
+        }
+
+        // Use cache to get or create PartialAggregateRowMerger
+        // This avoids repeated object creation and BitSet construction
+        CacheKey cacheKey = new CacheKey(schemaId, targetColumns);
+        return partialMergerCache.get(
+                cacheKey,
+                k -> {
+                    // Convert target column indices to column IDs
+                    List<Schema.Column> columns = schema.getColumns();
+                    Set<Integer> targetColumnIds = new HashSet<>();
+                    for (int colIdx : targetColumns) {
+                        targetColumnIds.add(columns.get(colIdx).getColumnId());
```

Comment:
Could you add a TODO on this that the `targetColumns` should already be column ids and not necessary to do the conversion.

cc @loserwang1024 , could you create issue for this? I think we have updated source projection pushdown to use column ids, but forgot to update sink upsert target columns.